File: signal-apps/src/signalcli.rs
use std::str;
use std::sync::Arc;

use async_process::{Child, Command, Stdio};
use async_trait::async_trait;
use futures_lite::{io::BufReader, prelude::*};
use tokio::sync::mpsc;

use crate::comm::{InternalSender, Receiver, Sender};

static SIGNALCLI_PATH: &str = "../signal-cli/build/install/signal-cli/bin/signal-cli";

pub struct SignalCliReciever {
    recv_chan: mpsc::Receiver<String>,
}

pub struct SignalCliSender {
    user: String,
}

pub struct SignalCliDaemon {
    daemon: Option<Child>,
    recvproc: Option<Child>,
    send_chan: Arc<mpsc::Sender<String>>,
}

impl SignalCliDaemon {
    pub fn new(user: &str) -> (Self, SignalCliReciever, SignalCliSender) {
        // TODO legit error handling
        let daemon = Some(
            Command::new(SIGNALCLI_PATH)
                .arg("daemon")
                .stdout(Stdio::null())
                .spawn()
                .unwrap(),
        );

        let (tx, rx) = mpsc::channel(10);

        let recv_chan = rx;
        let send_chan = Arc::new(tx);

        let mut recvproc = Command::new(SIGNALCLI_PATH)
            .arg("--dbus")
            .arg("--output=json")
            .arg("-u")
            .arg(user)
            .arg("receive")
            .arg("--timeout")
            .arg("-1") /* disable timeout */
            .stdout(Stdio::piped())
            .spawn()
            .unwrap();
        let recvout = recvproc.stdout.take().unwrap();
        let recvproc = Some(recvproc);

        {
            let send_chan = send_chan.clone();
            tokio::spawn(async move {
                let mut lines = BufReader::new(recvout).lines();
                // TODO max size for line?
                while let Some(Ok(line)) = lines.next().await {
                    send_chan
                        .try_send(line)
                        .expect("Sending received msg failed!");
                }
            });
        }

        (
            SignalCliDaemon {
                daemon,
                recvproc,
                send_chan,
            },
            SignalCliReciever { recv_chan },
            SignalCliSender {
                user: user.to_string(),
            },
        )
    }
}

impl InternalSender for SignalCliDaemon {
    fn insert_msg(&self, msg: &str) {
        self.send_chan
            .try_send(msg.to_string())
            .expect("Inserting message failed!");
    }

    fn stop(&mut self) {
        // TODO Maybe stop should actually be a custom drop?
        // investigate kill_on_drop in Command.
        eprintln!("terminating daemon");
        let mut dproc = self.daemon.take().unwrap();
        let mut recvproc = self.recvproc.take().unwrap();
        tokio::spawn(async move {
            dproc.kill().unwrap();
            dproc
                .status()
                .await
                .expect("failed to wait for daemon proc");

            recvproc.kill().unwrap();
            recvproc
                .status()
                .await
                .expect("failed to wait for recvproc");
            eprintln!("done terminating daemon");
        });
    }
}

#[async_trait]
impl Receiver for SignalCliReciever {
    async fn get_msg(&mut self) -> Option<String> {
        self.recv_chan.recv().await
    }
}

impl Sender for SignalCliSender {
    fn send(&self, dest: &str, msg: &str) {
        let dest = dest.to_string();
        let msg = msg.to_string();
        let user = self.user.clone();
        eprintln!("Starting send proc");
        tokio::spawn(async move {
            Command::new(SIGNALCLI_PATH)
                .args(&["--dbus", "-u", &user, "send", "-m", &msg, &dest])
                .stdout(Stdio::null())
                .output()
                .await
                .expect("Send failed!");
            eprintln!("Finished send proc");
        });
    }
}

File: prototype/signalcli.py
import json

import os
import subprocess
import threading

from queue import Queue
from typing import Callable

from sender import Sender
from receiver import Receiver

SIGNALCLI_PATH = "build/install/signal-cli/bin/signal-cli"
def do_chdir():
    os.chdir("signal-cli")

class SignalCliDaemon:
    def __init__(self, user: str) -> None:
        self.user = user
        daemon_cmd = [SIGNALCLI_PATH, 'daemon']
        self.daemon = subprocess.Popen(
            daemon_cmd,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            preexec_fn=do_chdir
        )

    def stop(self):
        print("Waiting for daemon to stop");
        self.daemon.terminate()
        self.daemon.wait()
        print("done Waiting for daemon to stop");


class SignalCliReceiver(Receiver):
    def __init__(self, daemon: SignalCliDaemon):
        self.daemon = daemon

        self.msgs = Queue()
        self.cbs = []

        recv_cmd = [
            SIGNALCLI_PATH,
            "--dbus",
            "--output=json",
            "-u",
            self.daemon.user,
            "receive",
            "--timeout",
            "-1", # disables timeout
        ]

        self.recv_proc = subprocess.Popen(
            recv_cmd, stdout=subprocess.PIPE, preexec_fn=do_chdir
        )

        self._stop_recv = False
        def recv_thread():
            while not self._stop_recv:
                msg = self.recv_proc.stdout.readline()
                if msg:
                    self.msgs.put(msg)
        self.recv_thread = threading.Thread(target=recv_thread)
        self.recv_thread.start()

        def cb_thread():
            while True:
                msg = self.msgs.get()
                if msg is None:
                    break
                try:
                    msg = json.loads(msg.decode())
                    for cb in self.cbs:
                        cb(msg)
                except:
                    print("Could not decode msg")
                    print(msg)
        self.cb_thread = threading.Thread(target=cb_thread)
        self.cb_thread.start()

    def stop(self) -> None:
        self.daemon.stop()

        self._stop_recv = True
        self.recv_proc.terminate()
        self.recv_proc.wait()
        self.recv_thread.join()

        self.msgs.put(None)
        self.cb_thread.join()

    def add_cb(self, on_msg_cb: Callable[[dict], None]) -> None:
        self.cbs.append(on_msg_cb)


class SignalCliSender(Sender):
    def __init__(self, daemon) -> None:
        self.daemon = daemon

    def send(self, dest: str, msg: str) -> None:
        send_cmd = [
            SIGNALCLI_PATH,
            "--dbus",
            "-u",
            self.daemon.user,
            "send",
            "-m"
        ]
        print("Sending msg", dest, msg)

        subprocess.check_call(send_cmd + [msg, dest], preexec_fn=do_chdir)
        print("done")
